### ISSUE TITLE
TestCafe goes into an infinite page reload loop in Edge. Browser gets disconnected and 'page loading' error screen is shown.

### ISSUE THREAD

**Issue by sijosyn** [DevExpress/testcafe#3028]
### What is your Test Scenario?
1 - Load this page url - https://aperture.synacor.net/404
2 - Script clicks the 'News' link. Takes to the News home page
3 - Go back to previous page
4 - Script clicks the 'Entertainment' link. Takes to the Entertainment home page
5 - Go back to previous page
6 - Script clicks the 'Sports' link. Takes to the Sports home page

### What is the Current behavior?
TestCafe goes into an infinite page reload loop when it tries to execute 4 & 5 in Edge browser. Won't go to step 6. After couple of page reload attempts, the Browser gets disconnected and a 'page loading' error screen is shown. (screenshot attached) . Same script works fine in Chrome on Windows.
 
### What is the Expected behavior?
Test should pass without any issues.
 
### What is your web application and your TestCafe test code?

 
Your website URL (or attach your complete example):

<details>
<summary>Your complete test code (or attach your test files):</summary>

 
```
import { Selector } from 'testcafe';

fixture `404 fixture`
    .page(`https://aperture.synacor.net/404`);

test('Clicking News, Entertainment and Sports link from 404 page opens the correct page', async t => {
    await t.click(Selector('ul').find('li').withText('News'));
    await t.expect(Selector('title').innerText).contains('News', 'Expected news in title.');
    await t.eval(() => window.history.back());
    await t.click(Selector('ul').find('li').withText('Entertainment'));
    await t.expect(Selector('title').innerText).contains('Entertainment', 'Expected entertainment in title.');
    await t.eval(() => window.history.back());
    await t.click(Selector('ul').find('li').withText('Sports'));
    await t.expect(Selector('title').innerText).contains('Sports', 'Expected sports in title.');
});
 
```
</details>

<details>
<summary>Your complete test report:</summary>

```
Please see screenshots attached below.
```
</details>

<details>
<summary>Screenshots:</summary>

![error](https://user-images.githubusercontent.com/29459281/47476030-51d57f00-d7ec-11e8-92e0-253f353eccdc.png)

![page_error](https://user-images.githubusercontent.com/29459281/47476042-69ad0300-d7ec-11e8-9c15-5c9d31f1d402.png)

</details>

### Steps to Reproduce:

1. Execute the provided script on Win 10 in Edge 16.16299
 
### Your Environment details:
 
* testcafe version:   0.23.0-alpha.1 . Also tried 0.22.0             
* node.js version:  8.11.1                
* command-line arguments:  testcafe edge test.js -e 
* browser name and version: Edge 16.16299 
* platform and version:  Win 10         
* other:

**Comment by VasilyStrelyaev** [comment]
Hi @sijosyn,
 
I've tried to reproduce the behavior you described but most of the time the test failed at the last assertion. Only once Edge tried to reload the tested page as you said.
 
At the same time, I've noticed that the website produces a lot of errors:
 
![image](https://user-images.githubusercontent.com/11459924/47500987-f9929180-d86c-11e8-97b7-351a03789179.png)
 
The markup is also incorrect (note the `<script>` tag after the closing `</html>` tag):
 
![image](https://user-images.githubusercontent.com/11459924/47501020-07481700-d86d-11e8-9bd6-14df642e7299.png)
 
Taking all this into consideration, it is hard to say whether the cause of the described behavior lies in TestCafe or the website itself. Could you please fix the website and try running this test again.

**Comment by sijosyn** [comment]
@VasilyStrelyaev Thanks for your prompt response. I installed the latest Windows OS updates on my machine and the Edge browser got updated to v17.17134.  I tried to run the same script in latest Edge v17 and the script runs fine without any issues. I'm unable to reproduce the browser disconnect issue. 

However, I noticed some performance issues when running the same script in different browsers on Win & Mac. Also the Safari 12 crashes with some strange error (see screenshot below).

Safari 12 Crash log screenshot:
<img width="1055" alt="screen shot 2018-10-25 at 6 14 39 pm" src="https://user-images.githubusercontent.com/29459281/47534186-6c5b3700-d884-11e8-932c-208cbe8b55db.png">

Time to took to execute script in different browsers:
- Safari 12.0.0 / Mac OS X 10.13.6 -  Browser crashes  (see log above)
- Firefox 62.0.0 / Mac OS X 10.13.0 - Took  **4m 45s** 
- Chrome 70.0.3538 / Mac OS X 10.13.6 - Took 32s
- Edge 17.17134.0 / Windows 10.0.0 - Took **1m 50s**
- Firefox 63.0.0 / Windows 10.0.0 - Took 53s
- Chrome 69.0.3497 / Windows 10.0.0. - Took 50s

Can you investigate why TestCafe takes very long time to execute such a small script? 

Regarding the page errors you mentioned, this is our demo page and we are aware of those page errors. I checked with our dev team about <script> tag after the closing </html> tag. This is the answer I got - "Most browsers self-correct such things. In chrome devtools it looks like it fixes it and auto moves it inside but you can see it when you view source."

**Comment by sijosyn** [comment]
@VasilyStrelyaev The requested debug information is provided below. Please review it. Regarding collecting performance statistics, I can open a separate ticket and provide all the details there with additional website URL's.

```javascript
Using locally installed version of TestCafe.
 Running tests in:
 - Safari 12.0.0 / Mac OS X 10.13.6

 404 fixture
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command     {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       "type": "click",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       "selector": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "type": "execute-selector",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "instantiationCallsiteName": "Selector",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "fnCode": "(function(){\n        var __f$=(function(){var selector=__dependencies$['selector'];var filter=__dependencies$['filter'];var filterNodes=__dependencies$['filterNodes'];var textRe=__dependencies$['textRe']; return (function () {\n/* eslint-disable no-undef */\nvar nodes = selector();\n\nif (!nodes.length) return null;\n\nreturn filterNodes(nodes, filter, document, void 0, textRe);\n/* eslint-enable no-undef */\n});})();;\n        return function(){\n            var args           = __dependencies$.boundArgs || arguments;\n            var selectorFilter = window['%testCafeSelectorFilter%'];\n            \n            var nodes = __f$.apply(this, args);\n            nodes     = selectorFilter.cast(nodes);\n            \n            if (!nodes.length && !selectorFilter.error)\n                selectorFilter.error = __dependencies$.apiInfo.apiFnID;\n\n            return selectorFilter.filter(nodes, __dependencies$.filterOptions, __dependencies$.apiInfo);\n        };\n     })();",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "args": [
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command           []
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         ],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "dependencies": [
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command           {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             "selector": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "@t": "Function",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "data": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "fnCode": "(function(){\n                var __f$=(function(){var selector=__dependencies$['selector'];var filter=__dependencies$['filter'];var filterNodes=__dependencies$['filterNodes'];var expandSelectorResults=__dependencies$['expandSelectorResults']; return (function () {\n/* eslint-disable no-undef */\nreturn expandSelectorResults(selector, function (node) {\n    if (typeof filter === 'string') {\n        return typeof node.querySelectorAll === 'function' ? node.querySelectorAll(filter) :null;\n    }\n\n    var results = [];\n\n    var visitNode = function visitNode(currentNode) {\n        var cnLength = currentNode.childNodes.length;\n\n        for (var i = 0; i < cnLength; i++) {\n            var child =  __get$(currentNode.childNodes,i);\n\n            results.push(child);\n\n            visitNode(child);\n        }\n    };\n\n    visitNode(node);\n\n    returnfilterNodes(results, filter, null, node);\n});\n/* eslint-enable no-undef */\n});})();;\n                return function(){\n                    var args           = __dependencies$.boundArgs|| arguments;\n                    var selectorFilter = window['%testCafeSelectorFilter%'];\n                    \n                    var nodes = __f$.apply(this, args);\nnodes     = selectorFilter.cast(nodes);\n                    \n                    if (!nodes.length && !selectorFilter.error)\n                        selectorFilter.error = __dependencies$.apiInfo.apiFnID;\n\n                    return selectorFilter.filter(nodes, __dependencies$.filterOptions, __dependencies$.apiInfo);\n                };\n             })();",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "dependencies": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   "selector": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "@t": "Function",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "data": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "fnCode": "(function(){\n   var __f$=(function(){return document.querySelectorAll(\"ul\");});;\n   return function(){\n       var args           = __dependencies$.boundArgs || arguments;\n       var selectorFilter = window['%testCafeSelectorFilter%'];\n       \n       var nodes = __f$.apply(this, args);\n       nodes     = selectorFilter.cast(nodes);\n       \n       if (!nodes.length && !selectorFilter.error)\n           selectorFilter.error = __dependencies$.apiInfo.apiFnID;\n\n       return selectorFilter.filter(nodes, __dependencies$.filterOptions, __dependencies$.apiInfo);\n   };\n})();",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "dependencies": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                         "filterOptions": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "filterVisible": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                             "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                             "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "filterHidden": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                             "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                             "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "counterMode": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                             "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                             "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "collectionMode": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "index": null
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                         },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                         "apiInfo": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "apiFnChain": [
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                             "Selector('ul')",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                             ".find('li')"
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           ],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "apiFnID": 0
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                         },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                         "boundArgs": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                         },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                         "customDOMProperties": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                         },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                         "customMethods": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                           "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                         }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   "filter": "li",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   "filterNodes": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "@t": "Function",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "data": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "fnCode": "(function(){ return (function (nodes, filter, querySelectorRoot, originNode) {for (var _len = arguments.length, filterArgs = Array(_len > 4 ? _len - 4 : 0), _key = 4; _key < _len; _key++) { __set$(filterArgs,_key-4,__get$(arguments,_key));}\n    if (typeof filter === 'number') {\n        var matchingNode = filter < 0 ?  __get$(nodes,nodes.length+filter) :  __get$(nodes,filter);\n\n        return matchingNode ? [matchingNode] : [];\n    }\n\n    var result = [];\n\n    if (typeof filter === 'string') {\n        // NOTE: we can search for elements only in document or element.\n        if (querySelectorRoot.nodeType !== 1 && querySelectorRoot.nodeType !== 9) return null;\n\n        var matching = querySelectorRoot.querySelectorAll(filter);\n        var matchingArr = [];\n\n        for (var i = 0; i < matching.length; i++) {matchingArr.push( __get$(matching,i));}\n\n        filter = function filter(node) {return matchingArr.indexOf(node) > -1;};\n    }\n\n    if (typeof filter === 'function') {\n        for (var j = 0; j < nodes.length; j++) {\n           if (filter.apply(undefined, [ __get$(nodes,j), j, originNode].concat(filterArgs))) result.push( __get$(nodes,j));\n        }\n    }\n\n    return result;\n});})();",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "dependencies": {}
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   "expandSelectorResults": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "@t": "Function",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "data": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "fnCode": "(function(){ return (function (selector, populateDerivativeNodes) {\n    var nodes = selector();\n\n    if (!nodes.length) return null;\n\n    var result = [];\n\n    for (var i = 0; i < nodes.length; i++) {\n        var derivativeNodes = populateDerivativeNodes( __get$(nodes,i));\n\n   if (derivativeNodes) {\n            for (var j = 0; j < derivativeNodes.length; j++) {\n                if (result.indexOf( __get$(derivativeNodes,j)) < 0) result.push( __get$(derivativeNodes,j));\n            }\n        }\n    }\n\n    return result;\n});})();",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "dependencies": {}
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   "filterOptions": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "filterVisible": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "filterHidden": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "counterMode": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "collectionMode": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "index": null
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   "apiInfo": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "apiFnChain": [
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       "Selector('ul')",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       ".find('li')",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                       ".withText('News')"
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     ],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "apiFnID": 1
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   "boundArgs": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   "customDOMProperties": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   "customMethods": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                     "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                   }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             "filter": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "@t": "Function",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "data": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "fnCode": "(function(){ return function selectorTextFilter(node, index, originNode, textFilter) {\n\n    function hasChildrenWithText(parentNode) {\n        var cnCount = parentNode.childNodes.length;\n\n        for (var i = 0; i < cnCount; i++) {\n            if (selectorTextFilter( __get$(parentNode.childNodes,i), index, originNode, textFilter)) return true;\n        }\n\n        return false;\n    }\n\n    function checkNodeText(text) {\n        if (textFilter instanceof RegExp) return textFilter.test(text);\n        return textFilter === text.trim();\n    }\n\n    // Element\n    if (node.nodeType === 1) {\n        var text = node.innerText;\n\n        // NOTE: In Firefox, <option> elements don't have `innerText`.\n        // So, we fallback to `textContent` in that case (see GH-861).\n        if (node.tagName.toLowerCase() === 'option') {\n            var textContent = node.textContent;\n\n            if (!text && textContent) text = textContent;\n        }\n\n        return checkNodeText(text);\n    }\n\n    // Document\n    if (node.nodeType === 9) {\n        // NOTE: latest version of Edge doesn't have `innerText` for `document`,\n        // `html` and `body`. So we check their children instead.\n        var head = node.querySelector('head');\n        var body = node.querySelector('body');\n\n        return hasChildrenWithText(head, textFilter) || hasChildrenWithText(body, textFilter);\n    }\n\n    // DocumentFragment\n    if(node.nodeType === 11) return hasChildrenWithText(node, textFilter);\n\n    return checkNodeText(node.textContent);\n};})();",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "dependencies": {}
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             "filterNodes": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "@t": "Function",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "data": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "fnCode": "(function(){ return (function (nodes, filter, querySelectorRoot, originNode) {for (var _len = arguments.length, filterArgs = Array(_len > 4 ? _len - 4 : 0), _key = 4; _key < _len; _key++) { __set$(filterArgs,_key-4,__get$(arguments,_key));}\n    if (typeof filter === 'number') {\n        var matchingNode = filter < 0 ?  __get$(nodes,nodes.length+filter) :  __get$(nodes,filter);\n\n        return matchingNode ? [matchingNode] : [];\n    }\n\n    var result = [];\n\n    if (typeof filter === 'string') {\n        // NOTE: we can search for elements only in document or element.\n        if (querySelectorRoot.nodeType !== 1 && querySelectorRoot.nodeType !== 9) return null;\n\n        var matching = querySelectorRoot.querySelectorAll(filter);\n        var matchingArr = [];\n\n        for (var i = 0; i < matching.length; i++) {matchingArr.push( __get$(matching,i));}\n\n        filter = function filter(node) {return matchingArr.indexOf(node) > -1;};\n    }\n\n    if (typeof filter === 'function') {\n        for (var j = 0; j < nodes.length; j++) {\n     if (filter.apply(undefined, [ __get$(nodes,j), j, originNode].concat(filterArgs))) result.push( __get$(nodes,j));\n        }\n    }\n\n    return result;\n});})();",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "dependencies": {}
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             "textRe": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "@t": "[[RegExp]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "data": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "src": "News",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "flags": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             "filterOptions": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "filterVisible": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "filterHidden": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "counterMode": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "collectionMode": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "index": null
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             "apiInfo": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "apiFnChain": [
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 "Selector('ul')",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 ".find('li')",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command                 ".withText('News')"
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               ],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "apiFnID": 2
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             "boundArgs": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             "customDOMProperties": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             "customMethods": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "@t": "[[undefined]]",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command               "data": ""
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command             }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command           }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         ],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "visibilityCheck": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "timeout": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "apiFnChain": [
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command           "Selector('ul')",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command           ".find('li')",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command           ".withText('News')"
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         ],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "index": 0
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       "options": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "speed": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "offsetX": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "offsetY": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "modifiers": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command           "ctrl": false,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command           "alt": false,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command           "shift": false,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command           "meta": false
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         "caretPos": null
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command     }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command  +0ms
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message     {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "cmd": "ready",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "status": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "id": "924",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "isCommandResult": false,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "executionError": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "pageError": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "resent": false,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "result": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "consoleMessages": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "log": [],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "info": [],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "warn": [],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "error": []
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "disableResending": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "allowRejecting": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "sessionId": "u36L5Rzmh"
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message     }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message  +965ms
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message     {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "cmd": "ready",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "status": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "id": "1204",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "isCommandResult": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "executionError": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "pageError": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "resent": false,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "result": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "consoleMessages": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "log": [],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "info": [],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "warn": [],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "error": []
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "disableResending": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "allowRejecting": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "sessionId": "u36L5Rzmh"
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message     }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message  +2s
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command     {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       "type": "execute-client-function",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       "instantiationCallsiteName": "eval",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       "fnCode": "(function(){ return (function () {return window.history.back();});})();",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       "args": [
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         []
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       ],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       "dependencies": [
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command         {}
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       ]
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command     }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command  +11ms
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message     {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "cmd": "ready",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "status": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "id": "23955.000000000004",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "isCommandResult": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "executionError": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "pageError": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "resent": false,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "result": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "consoleMessages": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "log": [],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "info": [],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "warn": [],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "error": []
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "disableResending": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "allowRejecting": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "sessionId": "u36L5Rzmh"
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message     }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message  +21s
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command     {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command       "type": "test-done"
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command     }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:command  +4ms
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message     {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "cmd": "ready",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "status": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "id": "23979",
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "isCommandResult": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "executionError": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "pageError": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "resent": false,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "result": null,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         "consoleMessages": {
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "log": [],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "info": [],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "warn": [],
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message           "error": []
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message         }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       },
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "disableResending": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "allowRejecting": true,
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message       "sessionId": "u36L5Rzmh"
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message     }
  testcafe:test-run:Safari 12.0.0 / Mac OS X 10.13.6:driver-message  +11ms
 ✖ Clicking News, Entertainment and Sports link from 404 page opens the correct page

   1) TypeError: Cannot read property '0' of null

      Browser: Safari 12.0.0 / Mac OS X 10.13.6

         266 |        this._handlePlainObject(val);
         267 |};
         268 |
         269 |DecodingTransformer.prototype.transform = function () {
         270 |    this.visitedRefs[0] = true;
       > 271 |    this._handleValue(this.references[0], this.references, 0);
         272 |
         273 |    return this.references[0];
         274 |};
         275 |
         276 |

         at DecodingTransformer.transform (/Users/sijocheeran/Documents/git/cube-e2e/node_modules/replicator/index.js:271:38)
         at module.exports.Replicator.decode (/Users/sijocheeran/Documents/git/cube-e2e/node_modules/replicator/index.js:559:24)
         at ClientFunctionBuilder.decode [as _processResult] (/Users/sijocheeran/Documents/git/cube-e2e/node_modules/testcafe/src/client-functions/client-function-builder.js:142:32)
         at _processResult (/Users/sijocheeran/Documents/git/cube-e2e/node_modules/testcafe/src/client-functions/client-function-builder.js:137:25)
         at $$testcafe_test_run$$u36L5Rzmh$$ (/Users/sijocheeran/Documents/git/cube-e2e/node_modules/testcafe/src/api/test-run-tracker.js:76:16)
         at $$testcafe_test_run$$u36L5Rzmh$$ (/Users/sijocheeran/Documents/git/cube-e2e/node_modules/testcafe/src/api/test-run-tracker.js:76:16)



 1/1 failed (24s)
```

**Comment by AndreyBelym** [comment]
The posted log shows that there is a problem with the page, because some driver messages which should appear in a normal logs are missing, but I can't find the reason. I'll try to use https://github.com/sijosyn/testcafe-syntests to reproduce the issue, and post an update here if I get the reason or need more information.

**Comment by sijosyn** [comment]
Thanks @AndreyBelym. I pushed some additional configs to my test repo. So if you would like to use a different website (they all share same codebase but with client specific logo, layout & style changes), just pass the config parameter in the command line. Same issue happens on those websites when tested in Safari.

>$ testcafe safari desktop --config=suddenlink -e
>$ testcafe safari desktop --config=mediacom -e

**Comment by sijosyn** [comment]
Issue is too old now and can't reproduce. Closing.
